File: 4136-concatenate-non-zero-digits-and-multiply-by-sum-ii/concatenate-non-zero-digits-and-multiply-by-sum-ii.cpp
const int mod = 1e9 + 7;
const int max_n = 100001;
long long pow10[max_n];

int init = []() {
    pow10[0] = 1;
    for(int i = 1; i < max_n; i++) {
        pow10[i] = (pow10[i-1] * 10) % mod;
    }
    return 0;
}();

class Solution {
public:
    vector<int> sumAndMultiply(string s, vector<vector<int>>& queries) {
        int n = s.size();
        vector<int> sum(n+1);
        vector<int> cnt(n+1);
        vector<long long> num(n+1);

        for(int i = 0; i < n; i++) {
            int d = s[i] - '0';
            sum[i+1] = sum[i] + d;
            cnt[i+1] = cnt[i] + (d > 0 ? 1 : 0);
            num[i+1] = d > 0 ? (num[i] * 10 + d) % mod : num[i];
        }

        vector<int> res(queries.size());
        int i = 0;
        for(const auto& q : queries) {
            int l = q[0];
            int r = q[1] + 1;
            int diff = cnt[r] - cnt[l];
            long long val = (num[r] - ((num[l]*pow10[diff])%mod)+mod) % mod;
            long long val_sum = sum[r] - sum[l];
            res[i++] = (int)((val * val_sum) % mod);
        }

        return res;
    }
};

File: 4136-concatenate-non-zero-digits-and-multiply-by-sum-ii/concatenate-non-zero-digits-and-multiply-by-sum-ii.java
class Solution {

    static int MOD = 1000000007;
    static int MAX = 100001;
    static long[] pow10 = new long[MAX];

    static {
        pow10[0] = 1;
        for(int i=1; i<MAX; i++) {
            pow10[i] = (pow10[i-1] * 10) % MOD;
        }
    }

    public int[] sumAndMultiply(String s, int[][] queries) {
        int n = s.length();
        int[] cnt = new int[n+1];
        int[] sum = new int[n+1];
        long[] num = new long[n+1];

        for(int i = 0; i < n; i++) {
            int d = s.charAt(i) - '0';
            sum[i+1] = d + sum[i];
            cnt[i+1] = d > 0 ? cnt[i]+1 : cnt[i];
            num[i+1] = d > 0 ? (num[i]* 10 + d) % MOD : num[i];
        }

        int[] res = new int[queries.length];
        int i = 0;
        for(int[] q : queries) {
            int l = q[0];
            int r = q[1] + 1;
            int diff = cnt[r] - cnt[l];
            long val = (num[r] - ((num[l] * pow10[diff]) % MOD) + MOD) % MOD;
            long val_sum = sum[r] - sum[l];
            res[i++] = (int)((val * val_sum) % MOD);
        }

        return res;
    }
}